PyArrowFileIO Implementation

samredai opened the pull request:
This PR adds the first FileIO implementation. It wraps the PyArrow [FileSystem](https://arrow.apache.org/docs/python/generated/pyarrow.fs.FileSystem.html#pyarrow.fs.FileSystem) class. The locations provided to `PyArrowInputFile` and `PyArrowOutputFile` are used to infer the correct filesystem by passing them to the [FileSystem.from_uri](https://arrow.apache.org/docs/python/generated/pyarrow.fs.FileSystem.html#pyarrow.fs.FileSystem.from_uri) method. The implementations of the `InputFile.open` and `OutputFile.create` methods both return a `pyarrow.lib.NativeFile` instance.

samredai commented:
@emkornfield

emkornfield commented on python/src/iceberg/io/base.py:
```diff
@@ -142,12 +158,25 @@ class FileIO(ABC):
+        """Get an OutputFile instance to write bytes to the file at the given location
+
+        Args:
+            location(str): The URI to the file
```
small nit, this isn't always a URI is it?  For local file systems it is just a file path which I don't think

samredai replied:
Updated! If a path with no scheme is provided, it's assumed as a path to a local file. Updated the tests as well and it feels much better not having to prepend `file:` :D

emkornfield commented on python/src/iceberg/io/pyarrow.py:
```diff
@@ -0,0 +1,213 @@
+        return input_file
+
+
+class PyArrowOutputFile(OutputFile):
```
For implementations, is there any reason to have 2 completely seperate classes?  Can't we have one that implements both?

danielcweeks replied:
I wouldn't necessarily want to change the abstraction of InputFile/OutputFile, but I like the multiple inheritance option that @samredai suggested as long as we don't have collisions that would cause issues.

samredai replied:
I consolidated `PyArrowInputFile` and `PyArrowOutputFile` into a single `PyArrowFile` class that inherits from both the `InputFile` and `OutputFile` base classes.

emkornfield commented on python/src/iceberg/io/pyarrow.py:
```diff
@@ -0,0 +1,213 @@
+        file_info = filesytem.get_file_info(path)
+        return False if file_info.type == FileType.NotFound else True
+
+    def open(self) -> NativeFile:
```
Yes, I would thinking of `-> InputStream`, I haven't coded much in python since typehints have come about, but it feels like generally for method overrides its not clear if you want specify a more specific type then the method that is getting overridden suggests.

samredai replied:
I just did a quick test and mypy will indeed check the overrides against the protocol. This is the right flexibility since if for example pyarrow one day decided to return `CustomFooFile` instead of `NativeFile`, mypy would raise the following **_only_** if it's missing the methods defined in the protocol:
```
Return type "CustomFooFile" of "open" incompatible with return type "InputStream" in supertype "InputFile"
```
So it feels like what we want--to only enforce adherence to the protocols.

samredai replied:
Updated the return types to be `InputStream` and `OutputStream`. 👍

kbendick commented on python/src/iceberg/io/pyarrow.py:
```diff
@@ -0,0 +1,215 @@
+        parsed_location = urlparse(location)  # Create a ParseResult from the uri
+
+        if parsed_location.scheme and parsed_location.scheme not in (
+            "file",
+            "mock",
+            "s3fs",
+            "hdfs",
+            "viewfs",
```
Nit: These are repeated multiple times. Could we just add them to a list or something somewhere and then reference that? The list could even be used in the raised exception too.

samredai replied:
Great idea, I'll add a `SUPPORTED_SCHEMES` variable at the module level.

samredai replied:
Updated

emkornfield commented on python/src/iceberg/io/pyarrow.py:
```diff
@@ -0,0 +1,215 @@
+            "hdfs",
+            "viewfs",
+        ):  # Validate that a uri is provided with a scheme of `file`
+            raise ValueError("PyArrowInputFile location must have a scheme of `file`, `mock`, `s3fs`, `hdfs`, or `viewfs`")
```
why is the value of scheme necessary?  For instance a new a GCS file should be present in the next version of Arrow?

samredai replied:
Good point about new schemes being supported in the future. We wouldn't have to come back here every time and update this check. The only thing is that it would be nice to fail early during an initialization of the wrapper class. Otherwise the user could do `input_file = PyArrowFile(location="foo://bar/baz.txt")` without any exception, and then later when running `input_file.open()` getting [here](https://github.com/apache/arrow/blob/master/cpp/src/arrow/filesystem/filesystem.cc#L722) and hitting `ArrowInvalid: Unrecognized filesystem type in URI: foo://bar/baz.txt`. I wonder if there's a light method available in pyarrow to just validate a URI, or maybe an array of supported schemes we could just import here and reference.

emkornfield replied:
I haven't tried it but I assume if you pass a URI for an unsupported arrow file system to from_uri it would raise an error?  Could you do that in the constructor?   I guess the downside is it might not be cheap.  I would have to check if there is any eager work like establishing a connection when constructing a file system.

samredai replied:
Yes from_uri seems to catch it. I'm thinking a bit more about this and I feel like it may also be worth avoiding the cost of instantiating a FileSystem for each call to the "open" or "create" method. Maybe we can use some kind of singleton wrapper that re-uses FileSystem instances (@CircArgs curious about your thoughts here). Is there any downside to keeping a FileSystem instance around for longer?

emkornfield replied:
I think doing the creation on the FileIO might make sense.  Or as you note having a cache which tries to find already existing equivelant FS instantiations.

samredai replied:
Based on the response on dev@arrow, it looks like caching here is reasonable. I updated the PR to do this and it actually streamlined a few things (removed the need for `parsed_location` to be a class attribute and made some patching in tests a bit easier where you can just override self._filesystem now).

The approach I took was adding a module level `_FILESYSTEM_INSTANCES` dictionary to use as a cache and a module level function `get_filesystem`. The function takes a location, pulls out the scheme, and then checks `_FILESYSTEM_INSTANCES` to if a filesystem already exists for that scheme. If the key doesn't exist, it instantiates a new filesystem using `FileSystem.from_uri(location)` and adds that filesystem instance to `_FILESYSTEM_INSTANCES`.

emkornfield commented on python/tests/io/test_pyarrow.py:
```diff
@@ -0,0 +1,130 @@
+
+
+@patch("iceberg.io.pyarrow.FileSystem")
+def test_pyarrow_violating_OutputStream_protocol(MockedFileSystem):
```
I see this pattern used elswhere but mixing camel cass and snake case doesn't look write to me.

samredai replied:
Yeah it does look odd, I'll update this.

samredai commented on python/src/iceberg/io/pyarrow.py:
```diff
@@ -0,0 +1,213 @@
+        filesytem, path = FileSystem.from_uri(self.location)  # Infer the proper filesystem
+        input_file = filesytem.open_input_file(path)
+        if not isinstance(input_file, InputStream):
+            raise TypeError("""Object returned from PyArrowInputFile.open does not match the InputStream protocol.""")
```
Added the type of the object to the error message. 👍

emkornfield commented on python/src/iceberg/io/pyarrow.py:
```diff
@@ -0,0 +1,178 @@
+        filesystem = _FILESYSTEM_INSTANCES[parsed_location.scheme]
+    else:  # Instantiate a filesystem and cache it by scheme
+        filesystem, _ = FileSystem.from_uri(location)  # Infer the proper filesystem
+        _FILESYSTEM_INSTANCES[parsed_location.scheme] = filesystem  # Cache the filesystem
```
I think we'll want to use WeakRef here?

samredai commented on python/src/iceberg/io/pyarrow.py:
```diff
@@ -0,0 +1,178 @@
+        else:
+            filesystem, _ = FileSystem.from_uri(os.path.abspath(location))
+            _FILESYSTEM_INSTANCES["local"] = filesystem
+    elif _FILESYSTEM_INSTANCES.get(parsed_location.scheme):  # Check for a cached filesystem
```
Hmm, I see, I just read Weston's response. I'll revert this commit and just do `from_uri` in the `PyArrowFile` constructor. Maybe we can then just wait to optimize that and think through some approaches at that time. How does that sound?

samredai replied:
I just realized urlparse actually parses the bucket name as `netloc` which makes sense since the bucket name is the authority in the URI. For example:
```py
from urllib.parse import urlparse
urlparse("s3://foobucket/test.txt")
# ParseResult(scheme='s3', netloc='foobucket', path='/test.txt', params='', query='', fragment='')
```
So we could cache for S3 only based on bucket like you suggested by using a tuple as the key. Something like this:
```py
parsed_location = urlparse(location)
if parsed_location.scheme == "s3":
  s3_fs_lookup = (parsed_location.scheme, parsed_location.netloc)
  if _FILESYSTEM_INSTANCES.get(s3_fs_lookup):
    filesystem = _FILESYSTEM_INSTANCES[s3_fs_lookup]
  else:
    filesystem, _ = FileSystem.from_uri(location)
    _FILESYSTEM_INSTANCES[s3_fs_lookup] = filesystem
    return filesystem

filesystem, _ = FileSystem.from_uri(location)
return filesystem
```

emkornfield replied:
I think maybe lets avoid the cache for now and benchmark stuff.  Based on the response I think in most cases it should hopefully be pretty cheap, but simpler is probably better to start.

rdblue commented on python/src/iceberg/io/pyarrow.py:
```diff
@@ -0,0 +1,178 @@
+from typing import Union
+from urllib.parse import urlparse
+
+from pyarrow.fs import FileSystem, FileType
```
Can we gracefully fail if pyarrow is not present? Or is that a matter of not importing this module?

emkornfield commented on python/src/iceberg/io/pyarrow.py:
```diff
@@ -0,0 +1,178 @@
+    @property
+    def exists(self) -> bool:
+        """Checks whether the file exists"""
+        file_info = self._filesystem.get_file_info(self.location)
```
this pattern looks strange to me, are we sure it works for non-local file systems?

samredai replied:
It does work for non-local file systems (at least for S3) and looks to be part of the FileSystem interface ([FileSystem.get_file_info](https://arrow.apache.org/docs/python/generated/pyarrow.fs.FileSystem.html#pyarrow.fs.FileSystem.get_file_info)). Also `type` is an attribute of the FileInfo interface ([FileInfo.type](https://arrow.apache.org/docs/python/generated/pyarrow.fs.FileInfo.html#pyarrow.fs.FileInfo.type)) that returns an enum value.

Although I did find one issue here. The full uri doesn't work but the `path` value returned from `from_uri` does, which is just the path part of the URI without the scheme or authority. I'm updating this now to store `self._filesystem, self._path = FileSystem.from_uri(location)` and using `self._path` here instead of `self.location`.

emkornfield commented on python/tests/io/test_pyarrow.py:
```diff
@@ -0,0 +1,126 @@
+    assert ("returned from PyArrowFile.open does not match the InputStream protocol.") in str(exc_info.value)
+
+
+def test_pyarrow_violating_output_stream_protocol():
```
it would be nice to add a test with something like minio to ensure we are getting the semantics we want on metastores.

samredai replied:
Do you mean like adding a test suite that launches a local minio server and uses it to mock s3?

emkornfield replied:
Thats what I was thinking.  The comment you mentioned above about having to adjust constructor above really covers my main concern in terms of consistent user experience.  It might be too heavy weight here.  I think for different implementations that access s3 (and other remote file systems) we might just need to accept that there might be quirks in behavior.

samredai commented:
@rdblue I believe I've addressed all of the comments here and it should be ready for a final review. Once we get to parts of the code that are using FileIO instances I think some aggressive testing there should suss out any bad assumptions here.

danielcweeks reviewed (approved):
+1 for me.  I found a couple strange behaviors bubbling up from the pyarrow FileSystem (like if you call `len(fin)` on a non-existing input file, it appears the c-bindings end up throwing an `OSError` as opposed to `FileNotFoundError` that I see in other areas of pyarrow FileSystem, but the message is still understandable.

That may never be a real issue and probably should be addressed in the underlying library, so everything here looks great.

kbendick reviewed (approved):
+1 this looks good to me. The Python refactor is really shaping up @samredai!

rdblue commented on python/src/iceberg/io/pyarrow.py:
```diff
@@ -0,0 +1,154 @@
+        file = self._filesystem.open_input_file(self._path)
+        return file.size()
+
+    @property
```
Why is this a property instead of a method? Since this will commonly make an RPC call and isn't cached, it seems like something we'd want people to think of as a method call.

samredai replied:
Good point, I switched this to a method here and also in the `InputFile` and `OutputFile` base classes.

rdblue commented on python/src/iceberg/io/pyarrow.py:
```diff
@@ -0,0 +1,154 @@
+
+    def __len__(self) -> int:
+        """Returns the total length of the file, in bytes"""
+        file = self._filesystem.open_input_file(self._path)
```
It looks like the size is available from `get_file_info`. Is it better to open the file to get the size? I would expect that `get_file_info` is better because it doesn't create an open file handle that needs to be closed.

If this is better for some reason, I'm fine using this. But we do need to close it if we don't switch to `get_file_info`.

samredai replied:
Definitely better to use `get_file_info`, updated!

rdblue commented on python/src/iceberg/io/pyarrow.py:
```diff
@@ -0,0 +1,154 @@
+        if not isinstance(input_file, InputStream):
+            raise TypeError(
+                f"Object of type {type(input_file)} returned from PyArrowFile.open does not match the InputStream protocol."
+            )
```
When do we expect this to happen? [`open_input_file`](https://arrow.apache.org/docs/python/generated/pyarrow.fs.FileSystem.html#pyarrow.fs.FileSystem.open_input_file) states that it returns `NativeFile`, which we know is an instance of the `InputStream` protocol. I think this check probably makes sense in tests, but I would not ever expect a failure here at runtime.

samredai replied:
You're right I think it's good enough to just include this in tests. I was thinking we would catch a change in the behavior of what pyarrow returns from `open_input_file`, but that would be something we catch in tests after updating the version or something along those lines, not at runtime.

samredai replied:
Fixed in [this](https://github.com/apache/iceberg/pull/4081/commits/7fa28c766c3dd5bd6702abcf4d85860c6bbb164c) commit.

rdblue commented on python/src/iceberg/io/base.py:
```diff
@@ -76,7 +76,15 @@ def close(self) -> None:
+    Attributes:
+        location(str): The URI or path to a local file for an InputFile instance
+        exists(bool): Whether the file exists or not
+    """
```
I just realized that we're missing a method for both `InputFile` and `OutputFile`: `close`! It's not needed to read or write, but it's really important!

samredai replied:
We have `close` as part of the `InputStream` and `OutputStream` protocol. I don't think we manage the file object's state in the `InputFile` and `OutputFile` classes, do we still need a `close` method in that case?

rdblue replied:
I must have missed it. I didn't see it on the protocol, but that's where I think it should be.

rdblue commented on python/src/iceberg/io/pyarrow.py:
```diff
@@ -0,0 +1,154 @@
+
+    @property
+    def exists(self) -> bool:
+        """Checks whether the file exists"""
```
Should this be more specific? This is actually checking whether the path exists, but it could be a directory. Maybe I'm splitting hairs though.

samredai replied:
Updated this to say "Checks whether the location exists" which makes more sense generally because the attribute is named `location`.

rdblue commented on python/src/iceberg/io/pyarrow.py:
```diff
@@ -0,0 +1,154 @@
+        """
+        if not overwrite and self.exists:
+            raise FileExistsError(
+                f"A file already exists at this location. If you would like to overwrite it, set `overwrite=True`: {self.location}"
```
In Iceberg, we try to stick to brief and direct error messages. Those typically look like this:
* `Cannot <what you tried>: <reason with context> (<optional help>)`
* `Failed to <low-level operation>: <reason with context>` (and wrap the underlying exception if there is one)
* `Invalid <something>: <what was invalid>`

Here, I'd recommend `"Cannot create file, already exists: {self.location}". I think it's fairly clear that you can pass an overwrite flag so I wouldn't mention it necessarily.

samredai replied:
Updated! Thanks, I'll make a note to keep the exception messages consistent going forward.

rdblue commented on python/src/iceberg/io/pyarrow.py:
```diff
@@ -0,0 +1,154 @@
+            raise FileExistsError(
+                f"A file already exists at this location. If you would like to overwrite it, set `overwrite=True`: {self.location}"
+            )
+        output_file = self._filesystem.open_output_stream(self._path)
```
If opening fails because the file was created concurrently, does it arrow raise `FileExistsError`?

samredai replied:
Did you mean if opening the file succeeds? For `open_output_stream`, the docstring says `If the target already exists, existing data is truncated.`. That would be a problem here where if `self.exists()` on the line above returns `False`, then a concurrent process creates and writes the files, then the `open_output_stream` here would successfully open the output stream but would truncate the contents. So it would effectively overwrite it even though `overwrite=False`.

@emkornfield, is there a way with pyarrow to open an output stream in a way that fails if the file already exists?

rdblue replied:
I see. So this is always an overwrite? That seems like a problem to me, but we may not be able to do anything about it besides have the existence check above. At a minimum, we should note that this can overwrite concurrently created files.

rdblue commented on python/src/iceberg/io/pyarrow.py:
```diff
@@ -0,0 +1,154 @@
+        Returns:
+            pyarrow.lib.NativeFile: A NativeFile instance for the file located at `self.location`
+        """
+        input_file = self._filesystem.open_input_file(self._path)
```
What is thrown when `open_input_file` is called for a file that doesn't exist?

samredai replied:
Ah I should spell that out in the docstring. The throw is not consistent here in that if it's a local file that doesn't exist, pyarrow throws a `FileNotFoundError` however if it's an S3 path it raises `OSError`. We can do a generic `Exception` catch and consistently raise a `FileNotFoundError` with the raw pyarrow error in the traceback. Or should we just defer to whatever exception pyarrow raises?

rdblue replied:
I think we should make the errors consistent for the caller and should document them in the Protocol.

rdblue commented on python/src/iceberg/io/pyarrow.py:
```diff
@@ -0,0 +1,154 @@
+        PyArrowFile class (as opposed to separate InputFile and OutputFile implementations), this method effectively returns
+        a copy of the same instance.
+        """
+        return PyArrowFile(self.location)
```
Can't this return `self`? I don't think that this class carries any state that needs to be reset.

samredai replied:
It definitely can, updated!

rdblue commented on python/src/iceberg/io/pyarrow.py:
```diff
@@ -0,0 +1,154 @@
+        try:
+            filesystem.delete_file(path)
+        except OSError:
+            raise FileNotFoundError(f"File could not be deleted because it does not exist: {str_path}")
```
Is it possible to pass the original exception as a cause? This assumes that any `OSError` means that the file doesn't exist, but that doesn't seem correct to me.

samredai replied:
I think this goes back to the inconsistency I described in my [other reply](https://github.com/apache/iceberg/pull/4081#discussion_r810664650). We could raise a more generic exception here and the user can see the more detailed pyarrow exception in the traceback:
```py
from pyarrow.fs import FileSystem

def delete(location):
    try:
        fs, path = FileSystem.from_uri(location)
        fs.delete_file(path)
    except Exception:
        raise RuntimeError(f"Cannot delete file: {location}")

delete("s3://foo/bar.txt")
```
output:
```py
OSError                                   Traceback (most recent call last)
<ipython-input-7-fd79114b19ff> in foo()
      3         fs, path = FileSystem.from_uri("s3://foo/bar.txt")
----> 4         fs.delete_file(path)
      5     except Exception:

/usr/local/lib/python3.9/site-packages/pyarrow/_fs.pyx in pyarrow._fs.FileSystem.delete_file()

/usr/local/lib/python3.9/site-packages/pyarrow/error.pxi in pyarrow.lib.check_status()

OSError: When getting information for key 'bar.txt' in bucket 'foo': AWS Error [code 15]: No response body.

During handling of the above exception, another exception occurred:

RuntimeError                              Traceback (most recent call last)
<ipython-input-8-c19b6d9633cf> in <module>
----> 1 foo()

<ipython-input-7-fd79114b19ff> in foo()
      4         fs.delete_file(path)
      5     except Exception:
----> 6         raise RuntimeError("Cannot delete file: s3://foo/bar.txt")
      7 
      8 

RuntimeError: Cannot delete file: s3://foo/bar.txt
```

rdblue replied:
Rather than RuntimeError, I'd use a more specific error type, but I like the try/except block to normalize this.

For OSError specifically, I think we also need to see what kind of OSError it is and allow it through if it isn't the one that means the file is missing.

rdblue replied:
Yeah, an [OSError](https://docs.python.org/3/library/exceptions.html#OSError) just passes the [C errno](https://man7.org/linux/man-pages/man3/errno.3.html) through. Here, you'll need to check for `ENOENT`, which is no such file or directory.

That probably also means that for the `open_output_stream` you can catch `OSError` and check `errno` for `EEXIST`, which indicates that the file exists.

samredai replied:
Digging some more I've found some more diverging behavior. For local files, `e.errno` is 2 as expected to represent `ENOENT` when the file does not exist.

For S3 though, there's no value set at all for `e.errno`. Instead:

- If the issue is access denied to the bucket, it hits this [ErrorToStatus](https://github.com/apache/arrow/blob/master/cpp/src/arrow/filesystem/s3_internal.h#L112) function that returns an `OSError` with nothing but a string description containing "AWS Error [code 15]" which corresponds to [ACCESS_DENIED](https://sdk.amazonaws.com/cpp/api/LATEST/_s3_errors_8h_source.html) in the aws cpp sdk.
- If the bucket access is fine but the resource doesn't exist, it hits this [PathNotFound](https://github.com/apache/arrow/blob/master/cpp/src/arrow/filesystem/s3fs.cc#L2143) error which returns an `OSError` with nothing but a string description containing "Path does not exist: foo".

I'm not enthusiastic about error handling based on substring checks, but the only thing I can think of here is checking for either `e.errno=2` or an error message containing "Path does not exist" in order to appropriately raise a `FileNotFoundError`. That feels fragile but should handle local and s3 files. I haven't checked the behavior for HDFS files but that could also be different.

samredai replied:
Updated the PR with this approach. The `exists`, `open`, and `create` methods now catch and appropriately raise errors such as `PermissionError`, `FileExistsError`, or `FileNotFoundError`.

For local files this is straightforward and checks `e.errno`, but for S3 errors, it checks the string response. If the error raised can't be determined, the underlying pyarrow error is just raised directly. Even when it's caught, the underlying pyarrow error is always available in the traceback.

I've also added a handful of tests to test for each of these scenarios and added docs for all of them in the `Raises:` section of the method docstrings.

rdblue reviewed (changes requested):
Thanks, @samredai, this is looking great! I noted a few things in the review but they should be pretty simple fixes.

rdblue commented on python/src/iceberg/io/base.py:
```diff
@@ -76,7 +76,15 @@ def close(self) -> None:
+
+    Attributes:
+        location(str): The URI or path to a local file for an InputFile instance
+        exists(bool): Whether the file exists or not
```
This is no longer an attribute/property right?

samredai replied:
Nope, I removed it here. I also removed filesystem since that's only used internally and not even set as a property.

rdblue commented on python/src/iceberg/io/pyarrow.py:
```diff
@@ -0,0 +1,179 @@
+        except OSError as e:
+            if e.errno == 13 or "AWS Error [code 15]" in str(e):
+                raise PermissionError(f"Cannot check if file exists, access denied: {self.location}")
+            raise  # pragma: no cover - If some other kind of OSError, raise the raw error
```
I would probably move this to `_file_info()` method that handles the exception and is shared by `__len__` and `exists`. Otherwise, you don't handle the exception in the `__len__` call to `get_file_info`.

samredai replied:
Updated!

rdblue commented on python/src/iceberg/io/pyarrow.py:
```diff
@@ -0,0 +1,179 @@
+            input_file = self._filesystem.open_input_file(self._path)
+        except OSError as e:
+            if e.errno == 2 or "Path does not exist" in str(e) or isinstance(e, FileNotFoundError):
+                raise FileNotFoundError(f"Cannot open file, does not exist: {self.location}")
```
Today I learned that Python automatically tracks causes from exceptions raised within except blocks. That's great; one of the annoying things about Java is that people forget to include the cause when they throw a new exception!

rdblue commented on python/src/iceberg/io/pyarrow.py:
```diff
@@ -0,0 +1,179 @@
+        """
+        try:
+            input_file = self._filesystem.open_input_file(self._path)
+        except OSError as e:
```
Rather than checking `isinstance(FileNotFoundError)` I think it would be better to throw the original. That way you don't have the longer error within an error.

This should work:

```python
try:
    ...
except FileNotFoundError:
    raise
except PermissionError:
    raise
except OSError as e:
    if e.errno == 2 or "Path does not exist" in str(e):
        raise FileNotFoundError(...)
    ...
```

samredai replied:
Updated!

rdblue commented on python/src/iceberg/io/pyarrow.py:
```diff
@@ -0,0 +1,188 @@
+        except OSError as e:
+            if e.errno == 2 or "Path does not exist" in str(e):
+                raise FileNotFoundError(f"Cannot delete file, does not exist: {location}")
+            elif e.errno == 13 or "AWS Error [code 15]" in str(e) or isinstance(e, PermissionError):
```
@samredai, here's one that wasn't updated to a separate except clause for `PermissionError`.

rdblue commented on python/src/iceberg/io/pyarrow.py:
```diff
@@ -0,0 +1,190 @@
+                an AWS error code 15
+        """
+        try:
+            file_info = self._filesystem.get_file_info(self._path)
```
Catch and raise FileNotFoundException?

samredai replied:
If the file does not exist pyarrow actually does not raise. It returns a FileInfo object with a type attribute of `FileType.NotFound`. So I added a check for that here and raise. 👍 

I then changed the `exists()` method to just do a try except where it catches the `FileNotFoundError` from `_file_info()`.

samredai commented on python/src/iceberg/io/pyarrow.py:
```diff
@@ -0,0 +1,190 @@
+            if not overwrite and self.exists() is True:
+                raise FileExistsError(f"Cannot create file, already exists: {self.location}")
+            output_file = self._filesystem.open_output_stream(self._path)
+        except PermissionError:
```
This goes back to the fact that `open_output_stream` just overwrites by default. This reminded me that I was supposed to add a note describing that so I updated this docstring with:
```
        Note:
            This retrieves a pyarrow NativeFile by opening an output stream. If overwrite is set to False,
            a check is first performed to verify that the file does not exist. This is not thread-safe and
            a possibility does exist that the file can be created by a concurrent process after the existence
            check yet before the output stream is created. In such a case, the default pyarrow behavior will
            truncate the contents of the existing file when opening the output stream.
```

samredai commented on python/src/iceberg/io/pyarrow.py:
```diff
@@ -0,0 +1,188 @@
+        filesystem, path = FileSystem.from_uri(str_path)  # Infer the proper filesystem
+        try:
+            filesystem.delete_file(path)
+        except OSError as e:
```
Yes it should since that's raised for local files, updated!

rdblue commented:
Nice work, @samredai!
